### PR TITLE
JIT: Run optRedundantBranches twice and try to compute more accurate doms

### LINKED ISSUE #70480: JIT: optRedundantBranches misses an opportunity
Extracted & simplified from https://github.com/dotnet/runtime/issues/36649#issuecomment-1149860134 to a separate issue 

```csharp
void PrintN(object o)
{
    if (o is int n)
        Console.WriteLine(n);
}
```
This simple method emits a lot of codegen:
```asm
; Method testout1:PrintN(System.Object):this
G_M58869_IG01:              ;; offset=0000H
       56                   push     rsi
       4883EC20             sub      rsp, 32
       488BF2               mov      rsi, rdx
G_M58869_IG02:              ;; offset=0008H
       488BD6               mov      rdx, rsi
       4885D2               test     rdx, rdx
       7442                 je       SHORT G_M58869_IG08
G_M58869_IG03:              ;; offset=0010H
       488B12               mov      rdx, qword ptr [rdx]
       48B9B86A32F4FA7F0000 mov      rcx, 0x7FFAF4326AB8      ; System.Int32
       483BD1               cmp      rdx, rcx
       7530                 jne      SHORT G_M58869_IG08
G_M58869_IG04:              ;; offset=0022H
       48B9B86A32F4FA7F0000 mov      rcx, 0x7FFAF4326AB8      ; System.Int32
       483BD1               cmp      rdx, rcx
       7413                 je       SHORT G_M58869_IG06
G_M58869_IG05:              ;; offset=0031H
       488BD6               mov      rdx, rsi
       48B9B86A32F4FA7F0000 mov      rcx, 0x7FFAF4326AB8      ; System.Int32
       FF15FC2BFDFF         call     [CORINFO_HELP_UNBOX]
G_M58869_IG06:              ;; offset=0044H
       8B4E08               mov      ecx, dword ptr [rsi+8]
G_M58869_IG07:              ;; offset=0047H
       4883C420             add      rsp, 32
       5E                   pop      rsi
       FF2586E73900         tail.jmp [System.Console:WriteLine(int)]
G_M58869_IG08:              ;; offset=0052H
       4883C420             add      rsp, 32
       5E                   pop      rsi
       C3                   ret      
; Total bytes of code: 88
```

image

From this flowgraph (built after "Redundant branch opts" phase) it's obvious that BB06 is dominated by BB02 with exactly the same condition which means BB06 can be folded (and BB07 becomes dead as well).

A not very TP-wise solution is basically run a sort of "UpdateFlowGraph" phase to drop unreachable blocks and recompute doms and then run `optRedundantBranches ` again

cc @AndyAyersMS

### PR DESCRIPTION
**Pull request by EgorBo**
Closes https://github.com/dotnet/runtime/issues/70480

```csharp
void Test(object o)
{
    if (o is int n)
        Console.WriteLine(n);
}
```
Asm diff for ^: https://www.diffchecker.com/8j7VGRuz

image

BB04 and BB07 are unreachable, because of "dead" BB04, BB06 reports BB01 as a dominator instead of BB02

Some nice [spmi diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1831382&view=ms.vss-build-web.run-extensions-tab)

### CODE REVIEW

**Comment by EgorBo** [comment]
PTAL @AndyAyersMS @dotnet/jit-contrib 
[Diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1831654&view=ms.vss-build-web.run-extensions-tab) up to 
```
Total bytes of delta: -54496 (-0.08 % of base)
```
for `libraries.pmi.windows.arm64.checked.mch`

File: src/coreclr/jit/flowgraph.cpp [hunk 1]
```diff
@@ -3109,6 +3109,48 @@ BasicBlock* Compiler::fgLastBBInMainFunction()
...
+            // can a block have a predEdge to itself?
+            return block->bbIDom;
+        }
+
+        // We check pred's count of InEdges - it's quite conservative.
+        // We, probably, could use fgReachable(fgFirstBb, pred) here to detect unreachable preds
+        if (predBlock->countOfInEdges() > 0)
+        {
+            lastReachablePred = predBlock;
+            reachablePreds++;
```
Comment (AndyAyersMS) [inline src/coreclr/jit/flowgraph.cpp:3142]: Seems like you can just bail out here once the count is > 1? Might save a little time.

File: src/coreclr/jit/redundantbranchopts.cpp [hunk 2]
```diff
@@ -62,6 +62,17 @@ PhaseStatus Compiler::optRedundantBranches()
         block->bbFlags &= ~BBF_VISITED;
     }
 
+    // We've made some changes so it makes sense to try
+    // one more time
+    if (visitor.madeChanges)
+    {
+        visitor.WalkTree();
```
Comment (AndyAyersMS) [inline src/coreclr/jit/redundantbranchopts.cpp:69]: I wonder if we can scope this down better somehow. Rewalking every block likely leads to a lot of redundant failures. We are paying nearly 1% in TP for this version.

One simple thing (which probably doesn't do much for TP) is that up in PostOrderVisit we can skip any block that has zero refs.

But I wonder if there's some other way to deduce which blocks likely would benefit (or conversely, likely would not benefit) from reconsideration.

There is fairly extensive literature on incremental dominator tree maintenance, for example
* [Dynamic Dominators and Low-High Orders in DAGs](https://drops.dagstuhl.de/opus/volltexte/2019/11171/pdf/LIPIcs-ESA-2019-50.pdf)
* [Incremental Computation of Dominator Trees](https://dl.acm.org/doi/pdf/10.1145/202529.202531)

so perhaps there's some clue in there about where revisiting might be fruitful.

Comment (AndyAyersMS) [reply src/coreclr/jit/redundantbranchopts.cpp:69]: Also, the dominator tree itself is not accurate. 

I think this just leads to missed opportunities as during the rewalk as we might visit blocks in the wrong order (we are trying for bottom-up). But I'm not 100% sure.

**Review by AndyAyersMS — COMMENTED** [review]
I like the idea, but it seems kind of costly.

**Comment by jakobbotsch** [comment]
Antigen has better coverage for loops than Fuzzlyn, so you may also want to consider running that (although I do see Fuzzlyn found something).

**Comment by AndyAyersMS** [comment]
Are the fuzzlyn/antigen failures real?

**Comment by EgorBo** [comment]
@AndyAyersMS PTAL, throughput regression is lowered from 1% to 0.01-0.1%, diff is slightly smaller but the pattern in question is handled. I plan a few more experiments here in future

File: src/coreclr/jit/flowgraph.cpp [hunk 3]
```diff
@@ -3116,6 +3116,53 @@ BasicBlock* Compiler::fgLastBBInMainFunction()
...
+    BasicBlock* lastReachablePred = nullptr;
+
+    // Check if we have unreachable preds
+    for (const flowList* predEdge : block->PredEdges())
+    {
+        BasicBlock* predBlock = predEdge->getBlock();
+        if (predBlock == block)
+        {
+            // can a block have a predEdge to itself?
+            return block->bbIDom;
```
Comment (AndyAyersMS) [inline src/coreclr/jit/flowgraph.cpp:3141]:
```suggestion
        if (predBlock == block)
        {
            continue;
```
You can just ignore these edges.

File: src/coreclr/jit/flowgraph.cpp [hunk 4]
```diff
@@ -3116,6 +3116,53 @@ BasicBlock* Compiler::fgLastBBInMainFunction()
...
+        }
+
+        // We check pred's count of InEdges - it's quite conservative.
+        // We, probably, could use fgReachable(fgFirstBb, pred) here to detect unreachable preds
+        if (predBlock->countOfInEdges() > 0)
+        {
+            lastReachablePred = predBlock;
+            reachablePreds++;
```
Comment (AndyAyersMS) [inline src/coreclr/jit/flowgraph.cpp:3149]: Nit: you don't need to count reachable preds here. If `lastReachablePred` is non-null, then there is more than one, the search fails, and you can set it to null and break.

Then below just check `lastReachablePred` is null or not.

File: src/coreclr/jit/redundantbranchopts.cpp [hunk 5]
```diff
@@ -47,7 +47,29 @@ PhaseStatus Compiler::optRedundantBranches()
...
                 madeChanges |= m_compiler->optRedundantRelop(block);
+
+                BasicBlock* bbNext = block->bbNext;
+                BasicBlock* bbJump = block->bbJumpDest;
+
                 madeChanges |= m_compiler->optRedundantBranch(block);
+
+                // It's possible that either bbNext or bbJump were unlinked and it's proven
+                // to be profitable to pay special attention to their successors.
+                if (madeChanges && (bbNext != nullptr) && (bbNext->countOfInEdges() == 0))
```
Comment (AndyAyersMS) [inline src/coreclr/jit/redundantbranchopts.cpp:58]: shouldn't `bbNext` always be non-null (since we're at a `BBJ_COND`) and likewise for `bbJump`?

**Review by AndyAyersMS — APPROVED** [review]
Happy that you found a very surgical way to check what might be affected.

**Pull request merged**